File: app/src/main/java/com/mmjang/ankihelper/data/ai/cache/AICacheRepository.java
package com.mmjang.ankihelper.data.ai.cache;

import android.util.Log;

import org.litepal.LitePal;
import org.litepal.LitePalDB;
import org.litepal.Operator;

import java.util.ArrayList;
import java.util.List;

public class AICacheRepository {
    
    // AI cache database will be initialized in MyApplication
    // This avoids potential conflicts with multiple LitePal.use() calls
    
    // AI Dictionary Cache methods
    public static List<AIDictionaryCache> getDictionaryCache(String word, long llmConfigId) {
        Log.d("AICacheRepository", "Querying cache for word: " + word + ", llmConfigId: " + llmConfigId);
        try {
            // Check if we can access the database
            Log.d("AICacheRepository", "Current database: " + LitePal.getDatabase().getPath());
            
            List<AIDictionaryCache> results = LitePal.where("hwd = ? and llmConfigId = ?", word, String.valueOf(llmConfigId))
                    .find(AIDictionaryCache.class);
            Log.d("AICacheRepository", "Found " + results.size() + " cached results for word: " + word);
            return results;
        } catch (Exception e) {
            Log.e("AICacheRepository", "Error querying cache for word: " + word, e);
            // Return empty list on error
            return new ArrayList<>();
        }
    }
    
    public static void saveDictionaryCache(AIDictionaryCache cache) {
        try {
            Log.d("AICacheRepository", "Saving cache entry for word: " + cache.getHwd());
            cache.save();
            Log.d("AICacheRepository", "Successfully saved cache entry for word: " + cache.getHwd());
        } catch (Exception e) {
            Log.e("AICacheRepository", "Error saving cache entry for word: " + cache.getHwd(), e);
        }
    }
    
    public static void clearExpiredDictionaryCache(long expirationTime) {
        LitePal.deleteAll(AIDictionaryCache.class, "timestamp < ?", String.valueOf(expirationTime));
    }
    
    public static int deleteOldestDictionaryCache(int count) {
        List<AIDictionaryCache> oldestRecords = LitePal.order("timestamp asc").limit(count).find(AIDictionaryCache.class);
        int deletedCount = 0;
        for (AIDictionaryCache cache : oldestRecords) {
            int result = cache.delete();
            if (result > 0) {
                deletedCount++;
            }
        }
        return deletedCount;
    }
    
    public static int deleteAllDictionaryCache() {
        return LitePal.deleteAll(AIDictionaryCache.class);
    }
    
    // AI Translator Cache methods
    public static AITranslatorCache getTranslatorCache(String sourceText, String sourceLanguage, 
                                                       String targetLanguage, long llmConfigId) {
        List<AITranslatorCache> results = LitePal.where("sourceText = ? and sourceLanguage = ? and targetLanguage = ? and llmConfigId = ?", 
                sourceText, sourceLanguage, targetLanguage, String.valueOf(llmConfigId))
                .find(AITranslatorCache.class);
        return results.isEmpty() ? null : results.get(0);
    }
    
    public static void saveTranslatorCache(AITranslatorCache cache) {
        cache.save();
    }
    
    public static void clearExpiredTranslatorCache(long expirationTime) {
        LitePal.deleteAll(AITranslatorCache.class, "timestamp < ?", String.valueOf(expirationTime));
    }
    
    public static int deleteOldestTranslatorCache(int count) {
        List<AITranslatorCache> oldestRecords = LitePal.order("timestamp asc").limit(count).find(AITranslatorCache.class);
        int deletedCount = 0;
        for (AITranslatorCache cache : oldestRecords) {
            int result = cache.delete();
            if (result > 0) {
                deletedCount++;
            }
        }
        return deletedCount;
    }
    
    public static int deleteAllTranslatorCache() {
        return LitePal.deleteAll(AITranslatorCache.class);
    }
}

File: app/src/main/java/com/mmjang/ankihelper/data/ai/cache/AICacheRepository.kt
package com.mmjang.ankihelper.data.ai.cache

import android.util.Log
import com.mmjang.ankihelper.MyApplication
import com.mmjang.ankihelper.data.database.AppDatabase
import com.mmjang.ankihelper.data.dict.CoroutineHelper

object AICacheRepository {

    // Helper method to get database instance
    private fun getDatabase(): AppDatabase {
        return AppDatabase.getInstance(MyApplication.getContext())
    }

    // AI Dictionary Cache methods
    @JvmStatic
    fun getDictionaryCache(word: String, llmConfigId: Long): List<AIDictionaryCache> {
        Log.d("AICacheRepository", "Querying cache for word: $word, llmConfigId: $llmConfigId")
        return try {
            val results = CoroutineHelper.executeBlocking {
                getDatabase().aiDictionaryCacheDao().getCacheByHeadwordAndLLM(word, llmConfigId)
            }
            Log.d("AICacheRepository", "Found ${results.size} cached results for word: $word")
            results
        } catch (e: Exception) {
            Log.e("AICacheRepository", "Error querying cache for word: $word", e)
            // Return empty list on error
            emptyList()
        }
    }

    @JvmStatic
    fun saveDictionaryCache(cache: AIDictionaryCache) {
        try {
            Log.d("AICacheRepository", "Saving cache entry for word: ${cache.hwd}")
            CoroutineHelper.executeBlocking {
                getDatabase().aiDictionaryCacheDao().insertCache(cache)
            }
            Log.d("AICacheRepository", "Successfully saved cache entry for word: ${cache.hwd}")
        } catch (e: Exception) {
            Log.e("AICacheRepository", "Error saving cache entry for word: ${cache.hwd}", e)
        }
    }

    @JvmStatic
    fun clearExpiredDictionaryCache(expirationTime: Long) {
        CoroutineHelper.executeBlocking {
            getDatabase().aiDictionaryCacheDao().deleteOldCache(expirationTime)
        }
    }

    @JvmStatic
    fun deleteOldestDictionaryCache(count: Int): Int {
        return CoroutineHelper.executeBlocking {
            getDatabase().aiDictionaryCacheDao().deleteOldestCache(count)
        }
    }

    @JvmStatic
    fun deleteAllDictionaryCache(): Int {
        CoroutineHelper.executeBlocking {
            getDatabase().aiDictionaryCacheDao().deleteAllCache()
        }
        return 1 // Room's delete returns void, so we return success indicator
    }

    @JvmStatic
    fun getDictionaryCacheCount(): Int {
        return CoroutineHelper.executeBlocking {
            getDatabase().aiDictionaryCacheDao().getCount()
        }
    }

    // AI Translator Cache methods
    @JvmStatic
    fun getTranslatorCache(
        sourceText: String,
        sourceLanguage: String,
        targetLanguage: String,
        llmConfigId: Long
    ): AITranslatorCache? {
        return CoroutineHelper.executeBlocking {
            getDatabase().aiTranslatorCacheDao().getCacheByTextAndLLM(
                sourceText, sourceLanguage, targetLanguage, llmConfigId
            )
        }
    }

    @JvmStatic
    fun saveTranslatorCache(cache: AITranslatorCache) {
        CoroutineHelper.executeBlocking {
            getDatabase().aiTranslatorCacheDao().insertCache(cache)
        }
    }

    @JvmStatic
    fun clearExpiredTranslatorCache(expirationTime: Long) {
        CoroutineHelper.executeBlocking {
            getDatabase().aiTranslatorCacheDao().deleteOldCache(expirationTime)
        }
    }

    @JvmStatic
    fun deleteOldestTranslatorCache(count: Int): Int {
        return CoroutineHelper.executeBlocking {
            getDatabase().aiTranslatorCacheDao().deleteOldestCache(count)
        }
    }

    @JvmStatic
    fun deleteAllTranslatorCache(): Int {
        CoroutineHelper.executeBlocking {
            getDatabase().aiTranslatorCacheDao().deleteAllCache()
        }
        return 1 // Room's delete returns void, so we return success indicator
    }

    @JvmStatic
    fun getTranslatorCacheCount(): Int {
        return CoroutineHelper.executeBlocking {
            getDatabase().aiTranslatorCacheDao().getCount()
        }
    }
}
